src/docfx/restore/Restore.cs

```diff
@@ -22,52 +45,78 @@ public static Task Run(string docsetPath, CommandLineOptions options, Report rep
                 report.Configure(docsetPath, config);
 
                 var restoredDirs = new HashSet<string>(PathUtility.PathComparer);
+                var workTreeMappings = new ConcurrentDictionary<string, Dictionary<string, string>>(PathUtility.PathComparer);
 
-                return ParallelUtility.ForEach(
-                    config.Dependencies.Values,
-                    (href, restoreChild) => RestoreDependentRepo(href, options, restoreChild, restoredDirs),
+                await ParallelUtility.ForEach(
+                    config.Dependencies.Values.Select(href => (docsetPath, href)),
+                    async (item, restoreChild) =>
+                    {
+                        var workTreeHead = await RestoreDependentRepo(item, options, restoreChild, restoredDirs);
+                        workTreeMappings.AddOrUpdate(
+                            item.docsetPath,
+                            new Dictionary<string, string> { { item.href, workTreeHead } },
+                            (key, value) =>
+                            {
+                                value[item.href] = workTreeHead;
```

Comment:
>value[item.href] [](start = 32, length = 16)

this looks thread unsafe

Reply:
seems the 'update' func can be executed more than once for the same key at the same time,  thanks for the reminder

---
In reply to: [200889864](https://github.com/dotnet/docfx/pull/3020#discussion_r200889864) [](ancestors = 200889864)